test/run-tests.py: Add an argument for running only failed tests.

stinos opened the pull request:
Implement the typical 're-run the failed tests' most test runners have for convenience.  Works by constructing the original test name back from the .exp files, so does require the current directory to be the same unless full test paths were used.

dpgeorge commented on tests/run-tests.py:
```diff
@@ -910,6 +915,11 @@ def main():
         help="delete the .exp and .out files from failed tests and exit",
     )
+    cmd_parser.add_argument(
+        "--run-failed",
```
maybe the option can be called `--run-failures`, to match the names of the other two options above

stinos replied:
Ok

dpgeorge commented on tests/run-tests.py:
```diff
@@ -974,7 +984,15 @@ def main():
         raise ValueError("target must be one of %s" % ", ".join(LOCAL_TARGETS + EXTERNAL_TARGETS))
 
-    if len(args.files) == 0:
+    if args.run_failed:
```
If `args.files` is specified, should that override `args.run_failed`?  I guess it doesn't really matter, but maybe for consistency `args.files` should take preference?

stinos replied:
With the wording 'run only failed' I'd say it takes precedence over passing files, but then indeed it's not clear that passing files as well will simply ignore them. But the same problem arises when doing it the other way around, then run_failed would silently be ignored. So if files should take precedence then either I change the doc on run-failures to state that it won't have effect if files are passed, or make it a hard error i.e. raise an ArgumentError (or something like that) if both are used?

dpgeorge replied:
Maybe just hard-fail if you pass both this arg and explicit files?

stinos replied:
Ok, seems most reasonable

dpgeorge commented on tests/run-tests.py:
```diff
@@ -85,6 +85,9 @@ def open(self, path, mode):
 """
 
+# Strings used in creating test basename, in the order they get applied.
+test_basename_strings = [(":", "&"), ("..", "#"), ("./", "%"), ("/", "$")]
```
This name mangling/unmangling is pretty tricky.  These symbols are not that easy to use in the shell (eg `#` is comment, `$` is variable prefix), and often I cat/diff these files explicitly, so need to type their names.

Maybe instead, there could be a `results/_failed_tests.txt` file which simply contains each failed test, one per line?  Or more general, `results/_manifest.txt` which contains all tests that were run, and the result (pass/FAIL).

stinos replied:
I'm indeed not too happy with this but changing it as you say needs some extra thought; I think such _manifests.txt would need to contain the test basename (for reference, to link it back to the .out and .exp files), and the path, and then the result. Would it be ok to instead use json like a dict of `{basename: (path, result)}`? Or you'd rather have a txt with each line `basename<sep>path<sep>result` but then with a `<sep>` which would be the ASCII record separator or something like that to avoid using a character which might occur in file paths?
Also if the file contains the result explicitly, should clean-failures then delete this file, or only the lines which failed?

dpgeorge replied:
I reckon it could be a `_results.json` file.  That should make it trivial to read/modify/write from Python.  (Using the name `manifest` might be too confusing with `manifest.py`...)

> Also if the file contains the result explicitly, should clean-failures then delete this file, or only the lines which failed?

Even if the file contained only the failures, it'd still need updating when re-running only these failures, because if some started to pass they'd need to be removed from the file.

Well, that could be done simply, by just always recreating the JSON file after `run-tests.py` finishes, with the list of failures.

I suggest to do the simplest thing that attains your original goal.

Might also be worth putting the `run-tests.py` command-line-arguments in the JSON.

stinos replied:
> I suggest to do the simplest thing that attains your original goal.

Ok. I did add command-line arguments (would also be useful to implement a 'do previous run again' command). Could add test count etc as well, just not sure how useful that is currently.

stinos replied:
Output will be like (this one is pretty printed):
```
{
    "args":
    {
        "target": "unix",
        "device": "/dev/ttyACM0",
        "baudrate": 115200,
        "user": "micro",
        "password": "python",
        "test_dirs":
        [
            "./basics"
        ],
        "result_dir": "C:/Users/stijn/projects/ghb/micropython/tests/results",
        "filters":
        [
            [
                "include",
                ".+"
            ]
        ],
        "write_exp": false,
        "list_tests": false,
        "emit": "bytecode",
        "heapsize": null,
        "via_mpy": false,
        "mpy_cross_flags": "-march=host",
        "keep_path": false,
        "jobs": 8,
        "files":
        [],
        "print_failures": false,
        "clean_failures": false,
        "run_failures": false
    },
    "failed_tests":
    [
        "./basics/builtin_help.py",
        "./basics/bytes_compare3.py"
    ]
}
```

stinos replied:
Hmm, would maybe be nicer to add the result writing in a separate function instead of just sticking everything in between the printing code. Or at least have it before the `print("{} tests performed...` line.

stinos committed:
tests/run-tests.py: Add an option for running only the failed tests.

Implement the typical 're-run the failed tests' most test runners have, for
convenience.  Accessible via the new --run-failures argument, and
implemented using a json file containing a list of the failed tests.

Signed-off-by: stijn <stijn@ignitron.net>

dpgeorge commented:
Thanks for this.  Tested using a bare-metal target and it works well.
